pkg/plugins/policies/meshtimeout/plugin/xds/configurer.go

```diff
@@ -94,19 +95,33 @@ func (c *DeprecatedListenerConfigurer) configureRequestTimeout(routeConfiguratio
 }
 
 func (c *DeprecatedListenerConfigurer) configureRequestHeadersTimeout(hcm *envoy_hcm.HttpConnectionManager) {
-	if conf := c.getConf(c.Subset); conf != nil {
+	// For backward, once a user upgrades from an older version we shouldn't set default timeouts.
+	// Refer to https://github.com/kumahq/kuma/issues/12033
+	deprecatedGetConf := c.legacyGetConf(c.Subset)
+	if deprecatedGetConf == nil {
+		return
+	}
+
+	if conf := c.getConf(c.Element); conf != nil {
 		hcm.RequestHeadersTimeout = toProtoDurationOrDefault(
 			pointer.Deref(conf.Http).RequestHeadersTimeout,
 			policies_defaults.DefaultRequestHeadersTimeout,
 		)
 	}
 }
 
-func (c *DeprecatedListenerConfigurer) getConf(subset rules.Subset) *api.Conf {
+func (c *DeprecatedListenerConfigurer) getConf(element rules.Element) *api.Conf {
+	if c.Rules == nil {
+		return &api.Conf{}
+	}
+	return rules.ComputeConf[api.Conf](c.Rules, element)
+}
+
+func (c *DeprecatedListenerConfigurer) legacyGetConf(subset rules.Subset) *api.Conf {
 	if c.Rules == nil {
 		return &api.Conf{}
 	}
-	return rules.ComputeConf[api.Conf](c.Rules, subset)
+	return rules.LegacyComputeConf[api.Conf](c.Rules, subset)
```

Comment:
We can use new Compute here, there is no difference when used with old outbound policies

Reply:
I found out why you're getting different results with Compute and LegacyCompute, the problem is in `Element#WithKeyValue` method:
```go
func (e Element) WithKeyValue(key, value string) Element {
	if e == nil {
		e = Element{}
	}
	
	e[key] = value
	return e
}
```
[you're modifying the underlying map `c.Element`](https://github.com/kumahq/kuma/blob/0485cd01ccb6efadda766185180ab0e3beecba04/pkg/plugins/policies/meshtimeout/plugin/xds/configurer.go#L80), but what we probably want is:
```go
func (e Element) WithKeyValue(key, value string) Element {
	c := maps.Clone(e)
	c[key] = value
	return c
}
```
If you fix `WithKeyValue` as I suggest then you won't see the difference between `ComputeConf` and `LegacyComputeConf`. So we can safely remove the legacy version